### ISSUE TITLE
`bool(factor(1))` evaluates to `False`

### ISSUE THREAD

**Issue by maxale** [sagemath/sage#38123]
### Steps To Reproduce

We have:
```
sage: bool(factor(1))
False
```

### Expected Behavior

As we typically identify factorizations with their values, and the factorization of 0 is not defined, conversion to `bool` of a factorization object should always result in `True`. 

In other words, we want `bool(F)` be the same as `bool(F.value())` for any factorization object `F`.

### Actual Behavior

Title says it.

### Environment

```markdown
- **OS**: Ubuntu 22.04.4 LTS
- **Sage Version**: 10.3
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by DaveWitteMorris** [comment]
-1 from me. I think of the output of `factor` as being a list, not a value, so the current behaviour seems correct.

**Comment by maxale** [comment]
Well, in the case of ambiguity, it makes better sense to raise an error about that than return a somewhat arbitrary undocumented value. Something like "ValueError: Boolean value of factor(1) is ambiguous." will do the job.

**Comment by DaveWitteMorris** [comment]
Documenting the value of `bool` may be a good idea, but I don't think we should raise an error, because I think the current behaviour is correct.  The output of `factor` has a `len`, and the python rule is that the boolean value of an object should be `False` when its `len` is 0.

**Comment by maxale** [comment]
@DaveWitteMorris:  I explained why it can be considered incorrect. Relation to `len` is a technical convention. From a mathematical point of view, I do not see why a factored form of an integer should be treated differently from an integer itself.

**Comment by mantepse** [comment]
I think I agree with @maxale.  We have code like
```
if x:
    do something
```
and I suspect that there will be some cases where we do not check the type of `x`.  I can imagine that passing a factored object instead of its value might then silently give wrong results.

Do you have an argument that this cannot happen?  I'm not sure how much a factorisation object is considered equivalent to its value.

**Comment by DaveWitteMorris** [comment]
i strongly disagree.  A factorization object should be thought of as a list of the factors, not as the value. (Perhaps the name of the method is misleading. For example, think of it as the `factors` method, which returns the factors.)  When iterating over the factors, it is important to know whether the iterator is empty.

**Comment by mantepse** [comment]
Why should a factorisation object be thought of as a list rather than it's value?

**Comment by DaveWitteMorris** [comment]
It has a `len` method (and you can iterate over it), so it's a container, not a value. As I said above, the python rule is that the boolean value of an object should be False when its`len` is 0. If a user wants the value, then they can take its value, but the object itself is list-like.  If we want to treat it as a value, then the `len` method should be removed, but I think that has a huge downside, because iterating over a factorization is what I think people usually do.  Otherwise, why bother factoring?

**Comment by maxale** [comment]
I suggested an alternative in raising an exception about the ambiguous conversion into `bool`. It will allow to keep the `len` method intact, and will encourage the user to make the condition more explicit depending on the underlying meaning - e.g. `x.value()!=0` or `len(x)!=0` or `x!=factor(1)` etc. Most importantly, it will eliminate the possibility of an unexpected result because of misinterpretation of the meaning `bool(factor(1))`.

**Comment by mantepse** [comment]
@DaveWitteMorris, OK, I wasn't aware of the `len => container` implication.  (`iterable` does not imply `container`, does it?  For example, polynomials have `list` but no `len`)

@maxale, unfortunately, that `bool` does not raise exceptions is another python philosophy issue, I'm afraid we'd break many user's expectations there.

**Comment by DaveWitteMorris** [comment]
I am not a python expert, but I can say: yes, only containers have a `len`. Quoting [Guido van Rossum](https://docs.python.org/3/faq/design.html#why-does-python-use-methods-for-some-functionality-e-g-list-index-but-functions-for-other-e-g-len-list): "When I read code that says len(x) I know that it is asking for the length of something. This tells me two things: the result is an integer, and the argument is some kind of container."

On the other hand, I do agree that it's a problem that methods can return incorrect results if they are expecting an integer but receive a factorization.

PS. It was suggested in a previous comment that a mathematician would not distinguish between a factored form of an integer and the integer itself, but either I disagree or I wouldn't use "factored form of an integer" to refer to the prime factorization. If a mathematician formally defined the prime factorization of an integer, I think that it would be some kind of sequence, and that it would be distinct from the integer itself. Perhaps partitions are a good analogy. A partition of n is a way of expressing n as a sum of natural numbers, but it is certainly distinct from the number n itself. For example, the partition 1 + 2 + 3 is an object with 3 parts, not just the number 6 (so the symbol "+" represents a separator, not the operation of addition).

**Comment by mantepse** [comment]
I think I am in favour of removing len for the factorisation object.

The situation with partitions is a lucky coincidence.

**Comment by grhkm21** [comment]
> I think I am in favour of removing len for the factorisation object.
> 
> The situation with partitions is a lucky coincidence.

-1 from me. len(factor(n)) is handy in many places. If `list` has `len` and `dict` has `len`, I don't see why `Factorization` shouldn't

**Comment by mantepse** [comment]
> > I think I am in favour of removing len for the factorisation object.
> > The situation with partitions is a lucky coincidence.
> 
> -1 from me. len(factor(n)) is handy in many places. If `list` has `len` and `dict` has `len`, I don't see why `Factorization` shouldn't

I found 8 occurrences of `len(factor(` in the sage library (6 in doctests, 2 in code).  This does not include constructions like
`f = factor(n); len(f)`, it would be slightly more complicated to find these.

However, my point is that I don't see why `Factorization` should be a container.  It is clear that, if it is a container, it must have `len`, but I think that allowing to compute with factorizations is a good argument for removing `len`.

Besides, I am frequently annoyed by factorization of `0` not being defined, is there a good reason for that?  One of the things I use sage a lot for is to factor the objects in a list, as in:
```
sage: [catalan_number(n).factor() for n in range(7)]
[1, 1, 2, 5, 2 * 7, 2 * 3 * 7, 2^2 * 3 * 11]
sage: [fibonacci(n).factor() for n in range(7)]
...
ArithmeticError: factorization of 0 is not defined
```
Instead I have to do:
```
sage: [f.factor() if f else 0 for n in range(7) if (f := fibonacci(n)) is not None]
[0, 1, 1, 2, 3, 5, 2^3]
```

**Comment by mantepse** [comment]
> Mathematically there is no set of primes that multiply to 0?

Oh yes.  But we could make the `Factorization` class slightly more general, so that `0` becomes an allowed value.  A new method, say `number_of_factors`, replacing `len`, would then throw an error.

I admit that I am not at all sure about this.

> We can make it return `None` if it's more convenient

I am not sure whether that would outweigh the inconvenience of changing a rather fundamental structure.

**Comment by DaveWitteMorris** [comment]
I think this discussion of 0.factor() is off-topic,so should move to a different ticket if it is going to continue. I think an error is the correct default behaviour for this, so I don't think it should be changed. However, I think it would be ok to add a keyword argument to let the user choose a different behaviour.

**Comment by mantepse** [comment]
> I think this discussion of 0.factor() is off-topic,so should move to a different ticket if it is going to continue. I think an error is the correct default behaviour for this, so I don't think it should be changed. However, I think it would be ok to add a keyword argument to let the user choose a different behaviour.

I think that the two issues are closely linked:

* the only reason I can see to make `Factorization` not a container, i.e., not have `len`, and therefore that `bool(factor(1))` evaluates to `True`, is to make it possible to compute with factored objects.  Computing with factored objects is something I would like to do frequently, so I think it is a valid question.
* However, if zero is not a `Factorization` object, computations with these don't work well anyway, and then it doesn't really make sense to me not to consider these objects as containers.

**Comment by saraedum** [comment]
Fyi, another Factorization oddity is at https://github.com/sagemath/sage/issues/38413.

**Comment by tscrim** [comment]
For the point about a mathematician identifying, e.g., an integer with its factorization, the point is a computer requires more explicitness than a mathematician allows. For example, the factorization as an integer versus as a rational number. We have a clear abuse that `(6/1).factor() == 6.factor()` because it is basically deemed too confusing for casual users, but we ended up with inconsistencies like
```
sage: R.<x> = QQ[]
sage: R(6/1).factor()
6
```
I think we need to make users more close attention to where particular objects live and perhaps be more explicit about things in their code.

It is also natural to iterate over factorization objects, which means there is a natural `len` too. Indeed, as noted a factorization is a multiset-like (container) object.
```
sage: list(6.factor())
[(2, 1), (3, 1)]
```

Because of these things, I am +1 on @DaveWitteMorris having `bool(factorization) == bool(len(factorization))`.
